### PR TITLE
Remove container lifecycle image dependency

### PR DESCRIPTION
**Pull request by Random-Liu**
Fixes https://github.com/containerd/cri/issues/990.

Test result without the fix:
```
--- FAIL: TestContainerLifecycleWithoutImageRef (1.66s)
    container_without_image_ref_test.go:29: Create a sandbox
    container_without_image_ref_test.go:42: Pull test image
    container_without_image_ref_test.go:49: Create test container
    container_without_image_ref_test.go:59: Remove test image
    container_without_image_ref_test.go:62: Container status should be running
        Error Trace:    container_without_image_ref_test.go:64
        Error:		Received unexpected error rpc error: code = Unknown desc = failed to get image "sha256:59788edf1f3e78cd0ebe6ce1446e9d10788225db3dedcfd1a59f764bad2b2690": does not exist
```

Test result with the fix:
```
=== RUN   TestContainerLifecycleWithoutImageRef
--- PASS: TestContainerLifecycleWithoutImageRef (1.74s)
    container_without_image_ref_test.go:29: Create a sandbox
    container_without_image_ref_test.go:42: Pull test image
    container_without_image_ref_test.go:49: Create test container
    container_without_image_ref_test.go:59: Remove test image
    container_without_image_ref_test.go:62: Container status should be running
    container_without_image_ref_test.go:67: Stop container
    container_without_image_ref_test.go:71: Container status should be exited
PASS
```

### CODE REVIEW

File: pkg/store/container/metadata.go [hunk 1]
```diff
@@ -56,6 +56,8 @@ type Metadata struct {
 	Config *runtime.ContainerConfig
 	// ImageRef is the reference of image used by the container.
 	ImageRef string
+	// StopSignal is the signal to send to stop the container.
+	StopSignal string
```
Comment (mikebrow) [inline pkg/store/container/metadata.go]: /s/StopSignal/Signal/

Comment (mikebrow) [inline pkg/store/container/metadata.go]: out of habit can we add new stuff to the end of the struct?

File: pkg/store/container/container_test.go [hunk 2]
```diff
@@ -78,7 +81,8 @@ func TestContainerStore(t *testing.T) {
 					Attempt: 1,
 				},
 			},
-			ImageRef: "TestImage-4abcd",
```
Comment (mikebrow) [inline pkg/store/container/container_test.go]: Needs a test that does not have a StopSignal.

Needs tests or TODO + issue for tests that use different signals.

OCI spec:
```
StopSignal string, OPTIONAL

The field contains the system call signal that will be sent to the container to exit. 
The signal can be a signal name in the format SIGNAME, for instance SIGKILL or SIGRTMIN+3.
```

**Comment by Random-Liu** [comment]
It is StopSignal. It matches the name in pic image config too

File: pkg/server/container_status.go [hunk 3]
```diff
@@ -43,17 +44,20 @@ func (c *criService) ContainerStatus(ctx context.Context, r *runtime.ContainerSt
...
-		// Based on current behavior of dockershim, this field should be
-		// image tag.
-		spec = &runtime.ImageSpec{Image: repoTags[0]}
-	}
-	if len(repoDigests) > 0 {
-		// Based on the CRI definition, this field will be consumed by user.
-		imageRef = repoDigests[0]
+		if err != store.ErrNotExist {
+			return nil, errors.Wrapf(err, "failed to get image %q", imageRef)
+		}
```
Comment (mikebrow) [inline pkg/server/container_status.go]: Let's cache the image.References in meta, we don't want to rely on it sticking around. Don't like the container status results changing / omitting image data cause the image ref has been removed from the store or worse the image has been updated / removed, but not for the container. Or at least we can add a TODO and open an issue to do so.

**Comment by mikebrow** [comment]
> It is StopSignal

yeah sorry hit the wrong button while reviewing I deleted that comment before finishing the review :-)

File: pkg/store/container/metadata.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/store/container/metadata.go]: Good point, will do.

File: pkg/store/container/container_test.go (hunk 2 above)
Comment (Random-Liu) [reply pkg/store/container/container_test.go]: Will add an integration test for this.

File: pkg/server/container_status.go (hunk 3 above)
Comment (Random-Liu) [reply pkg/server/container_status.go]: `image.References` is dynamically changed. I don't think we can cache it in container meta.

Comment (mikebrow) [reply pkg/server/container_status.go]: that's sort of my point, if the image that was used "changes" it's not what was used for the snapshot, thus the status information is not correct. I don't want to use the cache for anything but the status output.

image.References is just an array of strings.

Comment (Random-Liu) [reply pkg/server/container_status.go]: Are you ok with https://github.com/containerd/cri/issues/990#issuecomment-445323970?

File: pkg/store/container/metadata.go [hunk 4]
```diff
@@ -56,6 +56,8 @@ type Metadata struct {
 	Config *runtime.ContainerConfig
 	// ImageRef is the reference of image used by the container.
 	ImageRef string
+	// StopSignal is the signal to send to stop the container.
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/store/container/metadata.go]: Done

File: pkg/store/container/metadata.go (hunk 1 above)
Comment (Random-Liu) [reply pkg/store/container/metadata.go]: Done

File: pkg/store/container/container_test.go (hunk 2 above)
Comment (Random-Liu) [reply pkg/store/container/container_test.go]: Added TODO. Will send a follow up PR.

File: pkg/server/container_status.go (hunk 3 above)
Comment (mikebrow) [reply pkg/server/container_status.go]: yes, see responses on #990 

I still think we should cache the image.references at snapshot

Comment (mikebrow) [reply pkg/server/container_status.go]: nvm I see we do cache the original imageref.. that should be enough

Comment (Random-Liu) [reply pkg/server/container_status.go]: Are you OK with the change now? :)

**Pull request merged**